### PR TITLE
[WIP] Node 6 Support

### PR DESCRIPTION
**Pull request by ghusse**
I investigated on the Node 6 issue, and I found some interesting things. Unfortunately I did not fix all the issues (yet) but I think there are interesting things to know in this Pull Request.

Plus, if anyone has a clue about the origin of the regressions/issues left, I'm interested.

Things I noted:

1. Files in `/builtins` are processed by babel before being used. Which leads to strange behaviors, because `prelude.js` declares a global variable in strict environment.
2. Babel does not process `prelude.js` the same way in Node 6 and Node 8: an extra semi-column is added to the file in Node 6. Making it totally non-functional.

In this PR, I tried to fix these two problems by:

1. Make babel ignore files in `/builtins`, and use them only from `/src`
2. Change `prelude` to declare a function named `register`
3. Change `prelude` to write the result of this function call into `window.require` instead of using `require` without declaring it
4. Change the tests to create a global variable named `window` in the context of the vm
5. Change the way `JSPackager` combines different JS files, to create an IIFE around all the code and correctly call the function `register`
6. Call babel on test files (to be able to launch unit tests on Node 6)

I think all these remarks can be useful for people knowing more about the project than me (even if I learned A LOT about how parcel works in the process).

For the moment, more tests are passing on Node 6 than before my changes, but it seems that I introduced some regressions in Node 8. Any help is welcome.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -63,10 +63,11 @@
   "scripts": {
     "test": "cross-env NODE_ENV=test mocha",
     "format": "prettier --write './{src,bin,test}/**/*.{js,json,md}'",
-    "build": "babel src -d lib",
+    "build": "babel src -d lib --ignore src/builtins/*",
```
Comment (DeMoorJasper) [inline package.json]: Could u add this to .babelrc instead?

Comment (ghusse) [reply package.json]: Sure, thanks

**Comment by davidnagli** [comment]
Will this fix #40?

**Comment by ghusse** [comment]
Yes. But for the moment I need to investigate on regressions introduced by these changes.

**Comment by rondonjon** [comment]
@ghusse I really like that you took the time to enumerate and explain your changes. :+1:

**Comment by ghusse** [comment]
I found another issue with the current code.

Previous modifications caused some tests to fail on both node 8 and node 6. I did not understand why when I first created this PR. But since, I investigated and the results might interest you.

1/ Errors are ignored when running tests

First of all, `Bundler.bundle` totally hide errors during the bundle function. Which rends the debugging process very difficult.

I made a modification in this method to rethrow errors in test mode (I don't know why it should not be the case in other modes, but I did not want to introduce other regressions).

2/ Throwing errors revealed issues in some async functions

Throwing errors revealed the following error:

```
keyword super unexpected here
```

This error message seems to be caused by babel, not being able to correctly manage calls to `super` in `async` methods.

One workaround consisted of changing babel's configuration, the other consisted of using `ParentClass.prototype.functionName.call(this, [args])` instead of `super`.

I changed calls to super in async methods, and tests are now passing!

Tests are passing on Node 8, only 1 test is failing in node 6!

**Comment by ghusse** [comment]
I need help on this last issue: there is only 1 test failing on Node 6.

On node 8, both tests on `fs-destructuring` & `fs-destructuring-assignment` are passing.

On node 6, `fs-destructuring-assignment` is not passing, whereas `fs-destructuring` is almost similar, and is passing.

When debugging the code, it seems that, on passing tests, calls to `fs.readFileSync` are replaced by the actual content of the file.

But on node 6 + `fs-destructuring-assignment`, the call is not replaced (but code in `fs-destructuring` is correctly replaced "hello"). This issue in replacement causes the test to fail.

Is there anyone that can point me out the piece of code I should take a look at?

Thanks

**Comment by ghusse** [comment]
It seems that parcel uses babel to parse the JS Files, in order to replace calls to `readFileSync` with the actual file's content.

Babel does not parse files the same way with Node 6 and Node 8. Is there someone with more knowledge of the current parsing code, I'm interested by his/her insights on this subject.

**Comment by ghusse** [comment]
@devongovett? It seems that you wrote most of `src/visitor/fs`. Any clue?

**Comment by ghusse** [comment]
Up, I'd definitively need help on this last point. Is there someone knowing how Babel works internally and how it is used to replace calls to fs.readFileSync?

**Comment by devongovett** [comment]
> Babel does not parse files the same way with Node 6 and Node 8.

In what way? Like the resulting AST is different?

**Comment by ghusse** [comment]
Exactly, AST are different in both cases. I don't have the details right now, but i'll copy/paste them as soon as possible.

**Comment by ghusse** [comment]
On node8, in `visitors\fs.js`, the function `AssignmentExpression` is called twice:

I called console.log on `path.node`, here are the outputs in the console
```
// NODE 8
// first call with path.node =
Node {
  type: 'AssignmentExpression',
  start: 20,
  end: 50,
  loc:
   SourceLocation {
     start: Position { line: 2, column: 1 },
     end: Position { line: 2, column: 31 } },
  operator: '=',
  left:
   Node {
     type: 'ObjectPattern',
     start: 20,
     end: 34,
     loc: SourceLocation { start: [Object], end: [Object] },
     properties: [ [Object] ] },
  right:
   Node {
     type: 'CallExpression',
     start: 37,
     end: 50,
     loc: SourceLocation { start: [Object], end: [Object] },
     callee:
      Node {
        type: 'Identifier',
        start: 37,
        end: 44,
        loc: [Object],
        name: 'require' },
     arguments: [ [Object] ] },
  extra: { parenthesized: true, parenStart: 19 } }

// Second call on
Node {
  type: 'AssignmentExpression',
  start: 54,
  end: 116,
  loc:
   SourceLocation {
     start: Position { line: 3, column: 0 },
     end: Position { line: 3, column: 62 } },
  operator: '=',
  left:
   Node {
     type: 'MemberExpression',
     start: 54,
     end: 68,
     loc: SourceLocation { start: [Object], end: [Object] },
     object:
      Node {
        type: 'Identifier',
        start: 54,
        end: 60,
        loc: [Object],
        name: 'module' },
     property:
      Node {
        type: 'Identifier',
        start: 61,
        end: 68,
        loc: [Object],
        name: 'exports' },
     computed: false },
  right:
   Node {
     type: 'CallExpression',
     start: 71,
     end: 116,
     loc: SourceLocation { start: [Object], end: [Object] },
     callee:
      Node {
        type: 'Identifier',
        start: 71,
        end: 83,
        loc: [Object],
        name: 'readFileSync' },
     arguments: [ [Object], [Object] ] } }
```

On node 6, same test, same console.log

```
// NODE 6
// First call
{ type: 'AssignmentExpression',
  operator: '=',
  left:
   Node {
     type: 'Identifier',
     start: 21,
     end: 33,
     loc:
      SourceLocation {
        start: [Object],
        end: [Object],
        identifierName: 'readFileSync' },
     name: 'readFileSync',
     __clone: [Function: __clone] },
  right:
   { type: 'MemberExpression',
     object: { type: 'Identifier', name: '_require' },
     property:
      Node {
        type: 'Identifier',
        start: 21,
        end: 33,
        loc: [Object],
        name: 'readFileSync' },
     computed: false } }

// Second call
Node {
  type: 'AssignmentExpression',
  start: 54,
  end: 116,
  loc:
   SourceLocation {
     start: Position { line: 3, column: 0 },
     end: Position { line: 3, column: 62 } },
  operator: '=',
  left:
   Node {
     type: 'MemberExpression',
     start: 54,
     end: 68,
     loc: SourceLocation { start: [Object], end: [Object] },
     object:
      Node {
        type: 'Identifier',
        start: 54,
        end: 60,
        loc: [Object],
        name: 'module' },
     property:
      Node {
        type: 'Identifier',
        start: 61,
        end: 68,
        loc: [Object],
        name: 'exports' },
     computed: false },
  right:
   Node {
     type: 'CallExpression',
     start: 71,
     end: 116,
     loc: SourceLocation { start: [Object], end: [Object] },
     callee:
      Node {
        type: 'Identifier',
        start: 71,
        end: 83,
        loc: [Object],
        name: 'readFileSync' },
     arguments: [ [Object], [Object] ] } }
```

As you can see, the method `AssignmentExpression` is called twice in both cases, but on node 8 and node 6, input parameters are different.

<table>
<tr>
<th>Node 6 </th>
<th> Node 8</th>
</tr>
<tr><td><pre>{ type: 'AssignmentExpression',
  operator: '=',
  left:
   Node {
     type: 'Identifier',
     start: 21,
     end: 33,
     loc:
      SourceLocation {
        start: [Object],
        end: [Object],
        identifierName: 'readFileSync' },
     name: 'readFileSync',
     __clone: [Function: __clone] },
  right:
   { type: 'MemberExpression',
     object: { type: 'Identifier', name: '_require' },
     property:
      Node {
        type: 'Identifier',
        start: 21,
        end: 33,
        loc: [Object],
        name: 'readFileSync' },
     computed: false } }</pre></td>
<td><pre>Node {
  type: 'AssignmentExpression',
  start: 20,
  end: 50,
  loc:
   SourceLocation {
     start: Position { line: 2, column: 1 },
     end: Position { line: 2, column: 31 } },
  operator: '=',
  left:
   Node {
     type: 'ObjectPattern',
     start: 20,
     end: 34,
     loc: SourceLocation { start: [Object], end: [Object] },
     properties: [ [Object] ] },
  right:
   Node {
     type: 'CallExpression',
     start: 37,
     end: 50,
     loc: SourceLocation { start: [Object], end: [Object] },
     callee:
      Node {
        type: 'Identifier',
        start: 37,
        end: 44,
        loc: [Object],
        name: 'require' },
     arguments: [ [Object] ] },
  extra: { parenthesized: true, parenStart: 19 } }</pre></td></tr></table>

**Comment by ghusse** [comment]
@devongovett any idea?

**Pull request closed without merging**